app/models/saved_page.rb

```diff
@@ -1,9 +1,37 @@
 class SavedPage < ApplicationRecord
   belongs_to :oidc_user
 
+  validates :page_path, uniqueness: { scope: :oidc_user_id }, presence: true
+
+  validate :page_path_is_valid_path
+  validate :page_path_is_url_safe_string
+
   def to_hash
     {
       "page_path" => page_path,
     }
   end
+
+private
+
+  def page_path_is_url_safe_string
+    errors.add(:page_path, :url_unsafe, message: "must be a URL safe string") if is_url_unsafe_string?
+  end
+
+  def page_path_is_valid_path
+    errors.add(:page_path, :invalid_path, message: "must only include URL path") if is_invalid_path?
+  end
+
+  def is_invalid_path?
+    return true if page_path.blank?
+
+    uri = URI(CGI.unescape(page_path))
+    page_path != CGI.escape(uri.path)
```

Comment:
Why are we working with escaped strings?

If the request is something like

```
GET /saved-pages/%2Fsome%2Fpage%2Fon%2Fgov%2Fuk
```

Doesn't Rails automatically escape that for us in the controller to

```
params[:page_path] = "/some/page/on/gov/uk`
```

?

Manually (un)escaping strings is a place where errors could easily be introduced.

Reply:
Ah, I also see `is_url_unsafe_string?`

Right, we should be storing actual paths in our database, not escaped versions of them.